src/parcsr_mv/par_csr_matrix.c

```diff
@@ -785,6 +785,311 @@ hypre_ParCSRMatrixPrintIJ( const hypre_ParCSRMatrix *matrix,
+   HYPRE_Int             one = 1;
+   HYPRE_Int             myid, i, j, k;
+   HYPRE_BigInt          bigI, bigJ;
+   HYPRE_BigInt          ilower, iupper, jlower, jupper;
+   HYPRE_Complex         val;
+
+   /* MPI variables */
+   hypre_MPI_Comm_rank(comm, &myid);
+
+   /* Exit if trying to write from big-endian machine */
+   if ((*(char*)&one) == 0)
+   {
+      hypre_error_w_msg(HYPRE_ERROR_GENERIC, "Support to big-endian machines is incomplete!\n");
+      return hypre_error_flag;
+   }
+
+   /* Create temporary matrix on host memory if needed */
+   h_matrix = (hypre_GetActualMemLocation(memory_location) == hypre_MEMORY_DEVICE) ?
+              hypre_ParCSRMatrixClone_v2(matrix, 1, HYPRE_MEMORY_HOST) : matrix;
+
+   /* Matrix variables */
+   first_row_index = hypre_ParCSRMatrixFirstRowIndex(h_matrix);
+   first_col_diag  = hypre_ParCSRMatrixFirstColDiag(h_matrix);
+   diag            = hypre_ParCSRMatrixDiag(h_matrix);
+   offd            = hypre_ParCSRMatrixOffd(h_matrix);
+   col_map_offd    = hypre_ParCSRMatrixColMapOffd(h_matrix);
+   num_rows        = hypre_ParCSRMatrixNumRows(h_matrix);
+   row_starts      = hypre_ParCSRMatrixRowStarts(h_matrix);
+   col_starts      = hypre_ParCSRMatrixColStarts(h_matrix);
+
+   /* Diagonal matrix variables */
+   diag_nnz  = hypre_CSRMatrixNumNonzeros(diag);
+   diag_data = hypre_CSRMatrixData(diag);
+   diag_i    = hypre_CSRMatrixI(diag);
+   diag_j    = hypre_CSRMatrixJ(diag);
+
+   /* Off-diagonal matrix variables */
+   offd_nnz  = hypre_CSRMatrixNumNonzeros(offd);
+   offd_data = hypre_CSRMatrixData(offd);
+   offd_i    = hypre_CSRMatrixI(offd);
+   offd_j    = hypre_CSRMatrixJ(offd);
+
+   /* Set global matrix bounds */
+   ilower = row_starts[0] + (HYPRE_BigInt) base_i;
+   iupper = row_starts[1] + (HYPRE_BigInt) base_i - 1;
+   jlower = col_starts[0] + (HYPRE_BigInt) base_j;
+   jupper = col_starts[1] + (HYPRE_BigInt) base_j - 1;
+
+   /* Open binary file */
+   hypre_sprintf(new_filename, "%s.%05d.bin", filename, myid);
+   if ((fp = fopen(new_filename, "wb")) == NULL)
+   {
+      hypre_error_w_msg(HYPRE_ERROR_GENERIC, "Could not open output file\n");
+      return hypre_error_flag;
+   }
+
+   /*---------------------------------------------
+    * Write header (64 bytes)
+    *---------------------------------------------*/
+
+   count = 8;
+   header[0] = (hypre_uint64) sizeof(HYPRE_BigInt);
+   header[1] = (hypre_uint64) sizeof(HYPRE_Complex);
+   header[2] = (hypre_uint64) diag_nnz + offd_nnz;
+   header[3] = (hypre_uint64) ilower;
+   header[4] = (hypre_uint64) iupper;
+   header[5] = (hypre_uint64) jlower;
+   header[6] = (hypre_uint64) jupper;
+   header[7] = (hypre_uint64) 1;
+   if (fwrite((const void*) header, sizeof(hypre_uint64), count, fp) != count)
+   {
+      hypre_error_w_msg(HYPRE_ERROR_GENERIC, "Could not write all header entries\n");
+      return hypre_error_flag;
+   }
+
+   /* Allocate memory for buffers */
+   if (sizeof(HYPRE_BigInt) == sizeof(hypre_uint32))
```

Comment:
I like the way you check the sizes explicitly and store it in header information written in a fixed 64 bit format.

Reply:
That's the best solution I could think of for reaching portability across different platforms using the `ILP64`, `LP64`, `LLP64` and `LP32` models :)